opnode/l2/driver.go

```diff
@@ -2,88 +2,203 @@ package l2
+// parse log data for:
+//     event TransactionDeposited(
+//    	 address indexed from,
+//    	 address indexed to,
+//    	 uint256 value,
+//    	 uint256 gasLimit,
+//    	 bool isCreation,
+//    	 data data
+//     );
+func UnmarshalLogEvent(blockNum uint64, txIndex uint64, ev *types.Log) (*types.DepositTx, error) {
+	if len(ev.Topics) != 3 {
+		return nil, fmt.Errorf("expected 3 event topics (event identity, indexed from, indexed to)")
+	}
+	if ev.Topics[0] != DepositEventABIHash {
+		return nil, fmt.Errorf("invalid deposit event selector: %s, expected %s", ev.Topics[0], DepositEventABIHash)
+	}
+	if len(ev.Data) < 160 {
+		return nil, fmt.Errorf("deposit event data too small (%d bytes): %x", len(ev.Data), ev.Data)
+	}
 
-// sanity check that the receipts are consistent with the block data
-func (item *InputItem) CheckReceipts() bool {
-	hasher := trie.NewStackTrie(nil)
-	computed := types.DeriveSha(types.Receipts(item.Receipts), hasher)
-	return item.Block.ReceiptHash() == computed
-}
+	var dep types.DepositTx
+
+	dep.BlockHeight = blockNum
+	dep.TransactionIndex = txIndex
+
+	// indexed 0
+	dep.From = common.BytesToAddress(ev.Topics[1][12:])
+	// indexed 1
+	to := common.BytesToAddress(ev.Topics[2][12:])
+
+	// unindexed
+	// 0:32: value - big-endian
+	dep.Value = new(big.Int).SetBytes(ev.Data[0:32])
+	// 32:64: gas - big-endian
+	gas := new(big.Int).SetBytes(ev.Data[32:64])
+	if !gas.IsUint64() {
+		return nil, fmt.Errorf("bad gas value: %x", ev.Data[32:64])
+	}
+	dep.Gas = gas.Uint64()
+	// 64:96: isCreation - boolean, aligned with end.
+	// If False == 0 then it will create a contract using L2 account nonce to determine the created address.
+	if ev.Data[95] == 0 {
+		dep.To = &to
+	}
+	// 96:128: data offset
+	var dataOffset uint256.Int
+	dataOffset.SetBytes(ev.Data[96:128])
+	if dataOffset.Eq(uint256.NewInt(128)) {
+		return nil, fmt.Errorf("incorrect data offset: %v", dataOffset[0])
+	}
+
+	// 128:160: data length
+	var dataLen uint256.Int
+	dataLen.SetBytes(ev.Data[128:160])
+	if dataLen.Eq(uint256.NewInt(uint64(len(ev.Data) - 160))) {
+		return nil, fmt.Errorf("inconsitent data length: %v", dataLen[0])
+	}
+
+	// 160:...: data contents
+	dep.Data = ev.Data[160:]
+
+	// TODO: mint field
 
-type DepositTransaction struct {
-	Sender common.Address
-	// TODO some flag/bool for contract-deploy transaction. Or it can be *To with nil
-	Flags uint8
-	To    common.Address
-	Value uint256.Int
-	Data  []byte
+	return &dep, nil
 }
 
-func (dep *DepositTransaction) Decode(data []byte) (valid bool) {
-	// TODO
-	return false
+// CheckReceipts sanity checks that the receipts are consistent with the block data.
+func CheckReceipts(block *types.Block, receipts []*types.Receipt) bool {
+	hasher := trie.NewStackTrie(nil)
+	computed := types.DeriveSha(types.Receipts(receipts), hasher)
+	return block.ReceiptHash() == computed
 }
 
-type OpaqueTransaction []byte
+const L1InfoDepositIndex uint64 = 0xFFFF_FFFF_FFFF_FFFF
```

Comment:
At first this was called "L1 attributes deposits", then it was going to be "system deposit". I love a good naming bikeshed, but it would be nice to preserve a naming consistency, while we have the naming conversation elsewhere.

Reply:
Just picked something descriptive at that moment, I'll try to make naming more consistent with the remaining refactoring/polishing.